# Peak Annotation

In this workflow we use the peaks generated by macs2 in the previous workflow to perform de novo motif discovery and annotate each peak using using gene models for the reference genome and results of the motif analysis.

## Workflow

1. Filter peak intervals by peaks within negative control
2. Extract sequences surrounding peak summits
3. Generate random peak sequences distributed throughout the genome
4. Perform de novo motif discovery using DREME
5. Search for motifs within peaks using FIMO
6. Assess local motif enrichment using CENTRIMO
7. Compare de novo motifs to motif database using TOMTOM
8. Annotate filtered peak intervals using motif analysis results and gene models

## Merging and filtering peaks

```
cat merged_peaks.tsv | awk '$5 ~ /R1/ || $5 ~ /R3/ || $5 ~ /R4/{print $0}' | awk '{split($5,X,"-"); l=length(X); if($5~/R2/) {l=l-1} if($5~/R5/) {l=l-1} print $0,l}' > filtered_merged_peaks.tsv
```

Do this for each replicate

```
awk '$6 == 1{print $0}' filtered_merged_peaks.tsv | awk '{split($4,X,","); for (i in X) {print X[i]}}' > one_replicate_peaks.txt
```
