---
title: "Peak calling"
author: "Philipp Ross"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    theme: yeti
    highlight: default
    css: ../../assets/styles.css
---

```{r chunk-options, echo=F}
knitr::opts_chunk$set(echo = F, fig.align = "center", fig.width = 8, fig.height = 8, tidy = T, error = F, warning = F)
source("../../scripts/utils.R")
load_R_essentials()
sshhh("DT")
sshhh("GenomicFeatures")
sshhh("ChIPseeker")
gff <- "../../data/genomes/pfdd2.gff"
txdb <- makeTxDbFromGFF(gff)
```

## Average profiles

```{r load-files,echo=F}
# define files and file names
gfp_files <- list.files("../../data/peaks", "*_gfp_peaks.bed", full.names = T)
pnames <- toupper(
  str_replace(
    unlist(
      lapply(
        str_split(
          basename(gfp_files), pattern = "[.]"), 
        function(x) 
          {return(x[[1]])})),
    "_peaks", ""))

windows <- getPromoters(TxDb=txdb, upstream=5000, downstream=5000)
gfplist <- lapply(gfp_files, getTagMatrix, windows=windows)
names(gfplist) <- pnames
```

```{r plot-average-profiles, echo=F}
plotAvgProf(gfplist, xlim=c(-5000, 5000), conf=0.95, resample=100, facet="row", verbose=F)
```

## Annotations

```{r annotate-peaks,echo=F}
palist <- lapply(gfp_files, annotatePeak, TxDb=txdb, tssRegion=c(-1000,0), verbose=F)
names(palist) <- pnames
plotAnnoBar(palist)
plotDistToTSS(palist)
```

## Overlap of annotated genes

```{r overlap,echo=F}
genes <- lapply(palist, function(i) as.data.frame(i)$geneId)
vennplot(genes)
enrichPeakOverlap(queryPeak     = gfp_files[[4]],
                  targetPeak    = unlist(gfp_files[c(1,2,3,5)]), 
                  TxDb          = txdb, 
                  pAdjustMethod = "BH", 
                  nShuffle      = 50, 
                  chainFile     = NULL,
                  verbose       = FALSE)
```

```{r,eval=F,include=F,echo=F}
# Code to change the multi-intersect into a user-friendly format
bedtools multiinter -i ../../data/annotation/ap2i_r{1,2,3,4,5}_peak_intervals.narrowPeak -header -names R1 R2 R3 R4 R5 | sort -k1,1 -k2,2n > mintersect.tsv

tail -n +1 mintersect.tsv | awk 'BEGIN{i=0} {i+=1; print $1,$2,$3,"ipeak_"i,$4"_"$5}' | sed 's/,/-/g' > mintersect.bed

python ../../scripts/annotate_mintersect_peaks.py -m mintersect.bed -f ../../data/genomes/pfdd2_genes.gff -g ../../data/genomes/pfdd2.genome -n mintersect

df <- data.frame(import.gff3("../annotation/all_annotations_mintersect.gff"))

df <- df %>% select(-phase,-strand) %>% rowwise() %>% mutate(PBMMotifs = ifelse(length(PBMMotifs) == 0, ".", paste(PBMMotifs, collapse=" ")), DremeMotifs = ifelse(length(DremeMotifs) == 0, ".", paste(DremeMotifs, collapse=" ")), Replicates = paste(Replicates, collapse=" "))
> write_csv(df, "../annotation/all_annotations_mintersect.csv")
```
